Add `read_text_file()` function which detects and rejects invalid unicode characters

dennisschagt opened the pull request:
Fixes https://github.com/newsboat/newsboat/issues/844.

Current output when urls file contains an invalid unicode character:
![image](https://user-images.githubusercontent.com/4629607/101089582-a5af1680-35b5-11eb-88c7-8a65f7df6103.png)

I think it would be helpful to include the line number of the first line containing an invalid unicode character.

Minoru commented:
This is far better than a crash, so I'd merge this. However, I wonder if we can use iconv (the library) to convert the file from the locale's charset to UTF-8. With that, we won't have to explain to users why they need to keep their config in UTF-8 despite using e.g. KOI8-R for their locale.

Minoru commented:
Okay, lemme squash and merge #1203 then, it's unlikely anyone else will review it. And I'm also cool with not working on iconv right now.

Minoru commented:
@dennisschagt #1203 is merged, so you can rebase now.

dennisschagt commented:
> @dennisschagt #1203 is merged, so you can rebase now.

Thanks!

I added a commit which adds the current line number to error messages.
It adds quite some lines for something which might not be important.
@Minoru: Let me know if you want me to drop it (or if you have a better suggestion).

Minoru commented on src/utils.cpp:
```diff
@@ -924,6 +924,20 @@ std::string utils::getcwd()
 }
 
+bool utils::read_text_file(const std::string& filename,
+	std::vector<std::string>& contents, std::string& error_message)
```
Could this return `nonstd::expected<std::vector<std::string>, std::string>` instead? That'd be a bit more foolproof I think.

dennisschagt replied:
I somehow didn't notice that we have `nonstd::expected`.
Good idea, done in https://github.com/newsboat/newsboat/pull/1332/commits/fdb58e07c90d8e8addb14835c0c337f78f707faf.

Minoru commented on src/fileurlreader.cpp:
```diff
@@ -23,16 +23,15 @@ nonstd::optional<std::string> FileUrlReader::reload()
 	tags.clear();
 	alltags.clear();
 
-	std::fstream f;
```
Can we drop `#include <fstream>` now?

dennisschagt replied:
No, the `FileUrlReader::write_config()` method still uses an `std::fstream`:
https://github.com/newsboat/newsboat/blob/f2c16f01ebd839e6e7e32dcbd7d8759449bf20b2/src/fileurlreader.cpp#L63-L64

Minoru replied:
Ah, okay then. Thanks!

Minoru commented on rust/libnewsboat/src/utils.rs:
```diff
@@ -688,6 +691,37 @@ pub fn getcwd() -> Result<PathBuf, io::Error> {
+                reason.to_string()
+            ))
+        }
+    };
```
I think this could be replaced by `map_err`:

```rust
let file = File::open(filename)
    .map_err(|reason| fmt!(&gettext("Failed to open file (%s)"), reason.to_string()))?;
```

dennisschagt replied:
Nice improvement, thank you for the suggestion.
Applied in https://github.com/newsboat/newsboat/pull/1332/commits/bba28abe66187ce9504892440c7ec49b8b82ed6d.

Minoru commented on src/fileurlreader.cpp:
```diff
@@ -23,16 +23,15 @@ nonstd::optional<std::string> FileUrlReader::reload()
+	std::vector<std::string> lines;
+	std::string error_message;
+	if (!utils::read_text_file(filename, lines, error_message)) {
+		return strprintf::fmt(_("Error: Failed to read urls from file \"%s\" (%s)"),
```
I think this should say "URLs", because "urls" is just a file name.

dennisschagt replied:
Agreed, done in https://github.com/newsboat/newsboat/pull/1332/commits/c9ded7daddfccb85c9fc1c7ba0cb94f38be79638

Minoru reviewed (changes requested):
I think this is fine. One improvement I would request, though, is to *not* display the "no URLs configured" message if `UrlReader` failed — it's a bit confusing to have *two* error messages.

It would also be nice to display the offset into the line as well, but apparently Rust doesn't give us that info :(

dennisschagt commented:
> I think this is fine. One improvement I would request, though, is to not display the "no URLs configured" message if UrlReader failed — it's a bit confusing to have two error messages.

Just seeing I missed this. Will look into it.

Minoru reviewed (approved):
Looks great, thanks a lot for implementing this!

Minoru commented:
(I'm not merging this myself because I presume you'd like to squash those fixups first.)

dennisschagt commented:
> (I'm not merging this myself because I presume you'd like to squash those fixups first.)

Thanks, just rebased onto master and squashed the fixups.

Minoru commented:
Merged! Thanks a lot for the pull request.
